File: system_design/rate_limiter/python/go/rate_limiter_window.go
package main

import (
	"fmt"
	"time"

	"github.com/gomodule/redigo/redis"
)

type RateLimiterWindow struct {
	redis       redis.Conn
	windowSize  int
	maxRequests int
	prefix      string
}

func NewRateLimiterWindow(redisAddr string, windowSize int, maxRequests int) (*RateLimiterWindow, error) {
	conn, err := redis.Dial("tcp", redisAddr)
	if err != nil {
		return nil, err
	}
	return &RateLimiterWindow{
		redis:       conn,
		windowSize:  windowSize,
		maxRequests: maxRequests,
		prefix:      "rate_limit",
	}, nil
}

func (r *RateLimiterWindow) allowRequest(identifier string) (bool, string) {
	now := time.Now().Unix()
	key := fmt.Sprintf("%s:%s", r.prefix, identifier)
	windowStart := float64(now - r.windowSize)

	pipe := r.redis.pipeline()

	pipe.ZRangebyscore(key, windowStart, float64(now))
	pipe.ZCount(key, windowStart, float64(now))
	pipe.ZRange(key, 0, -1)
	pipe.ZAdd(key, map[string]float64{fmt.Sprintf("%d", now): float64(now)})
	pipe.ZRemRangeByScore(key, 0, windowStart-1)
	pipe.Expire(key, int64(r.windowSize))

	timestamps, currentCount, _, _, _, _, err := pipe.Execute()
	if err != nil {
		return false, fmt.Sprintf("Redis error: %v", err)
	}

	timeInWindow := float64(now - int(windowStart))
	rate := 0.0
	if timeInWindow > 0 {
		rate = float64(currentCount.(int64)) / timeInWindow
	}

	allowed := currentCount.(int64) < int64(r.maxRequests)
	message := fmt.Sprintf("Request %s (%d/%d requests in window, rate: %.2f/sec)",
		map[bool]string{true: "allowed", false: "denied"}[allowed],
		currentCount.(int64), r.maxRequests, rate)

	return allowed, message
}

func (r *RateLimiterWindow) getStatistics(identifier string) map[string]interface{} {
	key := fmt.Sprintf("%s:%s", r.prefix, identifier)
	windowStart := float64(time.Now().Unix() - r.windowSize)

	pipe := r.redis.pipeline()

	pipe.ZCount(key, windowStart, float64(time.Now().Unix()))
	pipe.ZRangebyscore(key, windowStart, float64(time.Now().Unix()))
	pipe.ZRange(key, 0, 0)
	pipe.ZRevRange(key, 0, 0)

	count, timestamps, oldest, newest, err := pipe.Execute()
	if err != nil {
		return map[string]interface{}{
			"identifier": identifier,
			"error":      "Failed to retrieve statistics",
		}
	}

	timestampsList, ok := timestamps.([]interface{})
	if !ok || len(timestampsList) == 0 {
		return map[string]interface{}{
			"identifier":           identifier,
			"allowed_requests":     0,
			"denied_requests":      0,
			"current_count":        0,
			"timestamp":            []float64{},
			"oldest_timestamp":     nil,
			"newest_timestamp":     nil,
			"window_usage_percent": 0.0,
		}
	}

	convertedTimestamps := make([]float64, len(timestampsList))
	for i, ts := range timestampsList {
		if t, ok := ts.([]interface{}); ok && len(t) > 1 {
			if val, ok := t[1].(float64); ok {
				convertedTimestamps[i] = val
			}
		}
	}

	allowed := int(count.(int64))
	denied := max(0, allowed-r.maxRequests)

	oldestTimestamp := oldest.([]interface{})
	newestTimestamp := newest.([]interface{})
	var oldestTS, newestTS *float64

	if len(oldestTimestamp) > 0 {
		if t, ok := oldestTimestamp[0].([]interface{}); ok && len(t) > 1 {
			if val, ok := t[1].(float64); ok {
				oldestTS = &val
			}
		}
	}

	if len(newestTimestamp) > 0 {
		if t, ok := newestTimestamp[0].([]interface{}); ok && len(t) > 1 {
			if val, ok := t[1].(float64); ok {
				newestTS = &val
			}
		}
	}

	return map[string]interface{}{
		"identifier":           identifier,
		"allowed_requests":     allowed,
		"denied_requests":      denied,
		"current_count":        allowed,
		"timestamp":            convertedTimestamps,
		"oldest_timestamp":     oldestTS,
		"newest_timestamp":     newestTS,
		"window_usage_percent": (float64(allowed) / float64(r.maxRequests)) * 100.0,
	}
}

func (r *RateLimiterWindow) reset(identifier string) bool {
	key := fmt.Sprintf("%s:%s", r.prefix, identifier)
	_, err := r.redis.Do("DEL", key)
	return err == nil
}

func (r *RateLimiterWindow) getRateLimitConfig(identifier string) map[string]interface{} {
	key := fmt.Sprintf("%s:%s", r.prefix, identifier)
	ttl, err := r.redis.Do("TTL", key)
	if err != nil {
		return map[string]interface{}{
			"identifier": identifier,
			"error":      "Failed to retrieve configuration",
		}
	}

	config := map[string]interface{}{
		"identifier":   identifier,
		"window_size":  r.windowSize,
		"max_requests": r.maxRequests,
		"current_ttl":  ttl,
		"rate":         0.0,
	}

	if ttl != nil {
		if rate, ok := ttl.(int64); ok && r.windowSize > 0 {
			config["rate"] = float64(r.maxRequests) / float64(r.windowSize)
		}
	}

	return config
}

func (r *RateLimiterWindow) Close() error {
	return r.redis.Close()
}

type RateLimiterWindowFixedSize struct {
	redis       redis.Conn
	windowSize  int
	maxRequests int
	prefix      string
}

func NewRateLimiterWindowFixedSize(redisAddr string, windowSize int, maxRequests int) (*RateLimiterWindowFixedSize, error) {
	conn, err := redis.Dial("tcp", redisAddr)
	if err != nil {
		return nil, err
	}
	return &RateLimiterWindowFixedSize{
		redis:       conn,
		windowSize:  windowSize,
		maxRequests: maxRequests,
		prefix:      "fixed_window",
	}, nil
}

func (r *RateLimiterWindowFixedSize) allowRequest(identifier string) (bool, string) {
	now := int64(time.Now().Unix())
	key := fmt.Sprintf("%s:%s:%d", r.prefix, identifier, now/r.windowSize)

	count, err := r.redis.Incr(key)
	if err != nil {
		return false, fmt.Sprintf("Redis error: %v", err)
	}

	allowed := count <= int64(r.maxRequests)
	message := fmt.Sprintf("Request %s (%d/%d requests in this window)",
		map[bool]string{true: "allowed", false: "denied"}[allowed],
		count, r.maxRequests)

	return allowed, message
}

func (r *RateLimiterWindowFixedSize) getStatistics(identifier string) map[string]interface{} {
	now := int64(time.Now().Unix())
	currentWindow := now / r.windowSize
	key := fmt.Sprintf("%s:%s:%d", r.prefix, identifier, currentWindow)

	count, err := r.redis.Do("GET", key)
	if err != nil {
		return map[string]interface{}{
			"identifier": identifier,
			"error":      "Failed to retrieve statistics",
		}
	}

	var countVal int64 = 0
	if count != nil {
		countVal = count.(int64)
	}

	return map[string]interface{}{
		"identifier":     identifier,
		"current_window": currentWindow,
		"current_count":  countVal,
		"window_size":    r.windowSize,
		"max_requests":   r.maxRequests,
		"current_ttl":    -2,
	}
}

func (r *RateLimiterWindowFixedSize) Close() error {
	return r.redis.Close()
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

File: system_design/rate_limiter/python/rate_limiter_window.py
"""
Sliding Window Rate Limiter using Redis

This implementation uses Redis to track request timestamps within a sliding window.
The algorithm maintains a record of all requests in the current time window
and calculates the actual rate based on the time elapsed within the window.

Time Complexity: O(n) where n is the number of requests in the window
Space Complexity: O(n) for storing request timestamps
"""

from typing import final
import time

try:
    import redis
except ImportError:
    redis = None


@final
class RateLimiterWindow:
    """
    Sliding Window Rate Limiter using Redis

    This implementation uses Redis to track request timestamps within a sliding window.
    The algorithm maintains a record of all requests in the current time window
    and calculates the actual rate based on the time elapsed within the window.

    Time Complexity: O(n) where n is the number of requests in the window
    Space Complexity: O(n) for storing request timestamps
    """

    def __init__(
        self, redis_client: redis.Redis, window_size: int = 60, max_requests: int = 100
    ):
        """
        Initialize the sliding window rate limiter.

        Args:
            redis_client: Redis client instance
            window_size: Size of the sliding window in seconds (default: 60)
            max_requests: Maximum number of requests allowed in the window (default: 100)
        """
        self.redis = redis_client
        self.window_size = window_size
        self.max_requests = max_requests
        self.prefix = "rate_limit"

    def allow_request(self, identifier: str) -> tuple[bool, str]:
        """
        Check if a request should be allowed based on the sliding window algorithm.

        Args:
            identifier: Unique identifier for the client (e.g., IP address, user ID)

        Returns:
            Tuple of (allowed, message)
        """
        now = time.time()
        key = f"{self.prefix}:{identifier}"
        window_start = now - self.window_size

        try:
            # Pipeline for atomic operations
            pipe = self.redis.pipeline()

            # 1. Get all request timestamps in the window
            pipe.zrangebyscore(key, window_start, now)

            # 2. Get the count of requests
            pipe.zcount(key, window_start, now)

            # 3. Get all request timestamps (for cleanup)
            pipe.zrange(key, 0, -1)

            # 4. Add the current request timestamp
            pipe.zadd(key, {str(now): now})

            # 5. Remove requests outside the window
            pipe.zremrangebyscore(key, 0, window_start - 1)

            # 6. Set expiration (TTL) to automatically clean up old keys
            pipe.expire(key, self.window_size)

            # Execute all commands
            timestamps, current_count, all_timestamps, added, removed, ttl = (
                pipe.execute()
            )

            # Calculate actual rate and time in window
            time_in_window = now - window_start
            rate = (
                (current_count / time_in_window)
                if time_in_window > 0
                else current_count
            )

            allowed = current_count < self.max_requests

            message = (
                f"Request {'allowed' if allowed else 'denied'} "
                f"({current_count}/{self.max_requests} requests in window, "
                f"rate: {rate:.2f}/sec)"
            )

            return allowed, message

        except redis.RedisError as e:
            # Handle Redis errors - consider the request denied to prevent blocking
            return False, f"Redis error: {str(e)}"

    def get_statistics(self, identifier: str) -> dict:
        """
        Get detailed statistics for a rate limiter instance.

        Args:
            identifier: Unique identifier for the client

        Returns:
            Dictionary containing statistics about the rate limiter state
        """
        key = f"{self.prefix}:{identifier}"
        window_start = time.time() - self.window_size

        try:
            pipe = self.redis.pipeline()

            # Get current count
            pipe.zcount(key, window_start, time.time())

            # Get requests in the last window
            pipe.zrangebyscore(key, window_start, time.time())

            # Get oldest request timestamp
            pipe.zrange(key, 0, 0, withscores=True)

            # Get newest request timestamp
            pipe.zrevrange(key, 0, 0, withscores=True)

            count, timestamps, oldest, newest = pipe.execute()

            if not timestamps:
                return {
                    "identifier": identifier,
                    "allowed_requests": 0,
                    "denied_requests": 0,
                    "current_count": 0,
                    "timestamp": [],
                    "oldest_timestamp": None,
                    "newest_timestamp": None,
                    "window_usage_percent": 0.0,
                }

            timestamps = [float(ts) for ts in timestamps]

            return {
                "identifier": identifier,
                "allowed_requests": int(count),
                "denied_requests": max(0, count - self.max_requests),
                "current_count": int(count),
                "timestamp": timestamps,
                "oldest_timestamp": float(oldest[0][1]) if oldest else None,
                "newest_timestamp": float(newest[0][1]) if newest else None,
                "window_usage_percent": (count / self.max_requests) * 100,
            }

        except redis.RedisError:
            return {
                "identifier": identifier,
                "error": "Failed to retrieve statistics",
            }

    def reset(self, identifier: str) -> bool:
        """
        Reset the rate limiter state for a specific identifier.

        Args:
            identifier: Unique identifier for the client

        Returns:
            True if successful, False otherwise
        """
        key = f"{self.prefix}:{identifier}"
        try:
            self.redis.delete(key)
            return True
        except redis.RedisError:
            return False

    def get_rate_limit_config(self, identifier: str) -> dict:
        """
        Get the current rate limit configuration for an identifier.

        Args:
            identifier: Unique identifier for the client

        Returns:
            Dictionary containing rate limit configuration
        """
        key = f"{self.prefix}:{identifier}"
        try:
            ttl = self.redis.ttl(key)
            return {
                "identifier": identifier,
                "window_size": self.window_size,
                "max_requests": self.max_requests,
                "current_ttl": ttl,
                "rate": (
                    (self.max_requests / self.window_size)
                    if self.window_size > 0
                    else 0
                ),
            }
        except redis.RedisError:
            return {
                "identifier": identifier,
                "error": "Failed to retrieve configuration",
            }


class RateLimiterWindowFixedSize:
    """
    Fixed Window Rate Limiter with Redis
    Similar to sliding window but uses fixed time intervals instead of a continuous window.
    Less accurate but uses less memory.
    """

    def __init__(
        self, redis_client: redis.Redis, window_size: int = 60, max_requests: int = 100
    ):
        """
        Initialize the fixed window rate limiter.

        Args:
            redis_client: Redis client instance
            window_size: Size of the time window in seconds (default: 60)
            max_requests: Maximum number of requests allowed in the window (default: 100)
        """
        self.redis = redis_client
        self.window_size = window_size
        self.max_requests = max_requests
        self.prefix = "fixed_window"

    def allow_request(self, identifier: str) -> tuple[bool, str]:
        """
        Check if a request should be allowed using fixed window algorithm.

        Args:
            identifier: Unique identifier for the client

        Returns:
            Tuple of (allowed, message)
        """
        now = int(time.time())
        key = f"{self.prefix}:{identifier}:{now // self.window_size}"

        try:
            count = self.redis.incr(key)

            # If this is the first request in this window, set expiration
            if count == 1:
                self.redis.expire(key, self.window_size + 1)

            allowed = count <= self.max_requests
            message = (
                f"Request {'allowed' if allowed else 'denied'} "
                f"({count}/{self.max_requests} requests in this window)"
            )

            return allowed, message

        except redis.RedisError as e:
            return False, f"Redis error: {str(e)}"

    def get_statistics(self, identifier: str) -> dict:
        """
        Get statistics for the fixed window rate limiter.

        Args:
            identifier: Unique identifier for the client

        Returns:
            Dictionary containing statistics
        """
        now = int(time.time())
        current_window = now // self.window_size
        key = f"{self.prefix}:{identifier}:{current_window}"

        try:
            count = self.redis.get(key)
            ttl = self.redis.ttl(key)

            return {
                "identifier": identifier,
                "current_window": current_window,
                "current_count": int(count) if count else 0,
                "window_size": self.window_size,
                "max_requests": self.max_requests,
                "current_ttl": ttl,
            }

        except redis.RedisError:
            return {
                "identifier": identifier,
                "error": "Failed to retrieve statistics",
            }
